PLUGIN-9 Include Metadata Annotations. Complex data types support

vladhlinskiy opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-15761
WIKI: https://wiki.cask.co/display/CE/SAP+ECC+Plugins

In the scope of this PR:
* Added support of the "Include Metadata Annotations" configuration option for both OData V2 and V4
* Nested annotations and annotations with external targeting are supported
* Added support of [Record](https://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/csprd05/odata-csdl-xml-v4.01-csprd05.html#sec_Record) expression annotations.
* Added support of the complex data types for both OData V2 and V4
* Added support of the OData4 `TypeDefinition` and `EnumType`

Currently, there is no support of other expression annotations, since Olingo V4 client does not parse expression annotations correctly: https://issues.apache.org/jira/browse/OLINGO-1403 .

vhlinskyi committed:
CDAP-15761 Support of the 'Include Metadata Annotations' configuration property added

vhlinskyi committed:
CDAP-15761 OData 4 'Include Metadata Annotations' configuration property

vhlinskyi committed:
CDAP-15761 Use CSDL model classes instead of Edm

vhlinskyi committed:
CDAP-15761 Support of nested annotations and annotations with external targeting added

vhlinskyi committed:
CDAP-15761 Complex data types support

vhlinskyi committed:
CDAP-15761 OData V2 complex data types support

vladhlinskiy commented:
@albertshau, could you, please, take a look?

albertshau commented on docs/SapEcc-batchsource.md:
```diff
@@ -32,6 +32,9 @@ see [OData URL components].
 https://www.odata.org/documentation/odata-version-3-0/url-conventions/
 
+**Include Metadata Annotations:** Whether the plugin should read SAP metadata annotations and include them to each
+record.
```
Should include some information about how this affects what gets output. Does it mean the annotation becomes an output field? If so, what is it's schema?

An example here may help make it clear

vladhlinskiy replied:
Changed to:

**Include Metadata Annotations:** Whether the plugin should read SAP metadata annotations and include them to each
record. In this case, each property mapped to a record with exactly two fields:
- value - for OData property value.
- metadata-annotations - metadata annotations record.

OData V4 metadata annotations mapped to a record with fields that correspond to the property attributes.
For example, for property, annotated as:
```
<Property Name="BuyerName" Type="Edm.String" Nullable="false" MaxLength="80" sap:label="Company"
                  sap:creatable="false" sap:updatable="false" sap:sortable="false" sap:filterable="false"/>
```
an entry with property value `true` will be read as the following record:
```
value: true
metadata-annotations:
  creatable: false
  updatable: false
  sortable: false
  filterable: false
```

OData V4 metadata annotations mapped to a record with the following fields:
- term - a simple identifier, such as "UI.DisplayName" or "Core.Description", etc.
- qualifier - a term can be applied multiple times to the same model element by providing a qualifier to distinguish
 the annotations.
- expression - record that corresponds to a constant expression or a dynamic expression.
- annotations - record that corresponds to nested annotations.

For example, for property, annotated as:
```
<Property Name="RecordNestedAnnotated" Type="Edm.Boolean" Nullable="false">
  <Annotation Term="Core.Description">
    <Annotation Term="Core.Description" Qualifier="nested" String="Nested annotation">
    <Record>
      <Annotation Term="Core.Description" Qualifier="onrecord" String="Annotation on record" />
      <PropertyValue Property="GivenName" Path="String"/>
      <PropertyValue Property="Age" Path="Byte"/>
    </Record>
  </Annotation>
</Property>
```
an entry with property value `true` will be read as the following record:
```
value: true
metadata-annotations:
 core_description:
  term: "Core.Description"
  qualifier: null
  expression:
    name: "Record"
    type: null
    propertyValues:
      GivenName:
        name: "Path"
        value: "SomeProperty1"
      Age:
        name: "Path"
        value: "SomeProperty2"
    annotations:
      onrecord_core_description:
        term: "Core.Description"
        qualifier: "onrecord"
        expression:
          name: "String"
          value: "Annotation on record"
  annotations:
    nested_core_description:
      term: "Core.Description"
      qualifier: "nested"
      expression:
        name: "String"
        value: "Nested annotation"
```

For more information, see [OData V4 Annotations].

[OData V4 Annotations]:
https://docs.oasis-open.org/odata/odata-csdl-xml/v4.01/csprd05/odata-csdl-xml-v4.01-csprd05.html#sec_Annotation

albertshau commented on src/main/java/io/cdap/plugin/sap/SapODataSource.java:
```diff
@@ -129,86 +152,322 @@ public Schema getSchema() {
+    return SapODataConstants.Annotation.schema(recordName, expressionSchema, annotationsSchema);
+  }
+
+  private @Nullable Schema nestedAnnotationsSchema(String annotationName, Map<String, OData4Annotation> annotations) {
```
style: nullable annotations go on top of the method

```
@Nullable
private Schema ...
```

vladhlinskiy replied:
Fixed.

albertshau commented on src/main/java/io/cdap/plugin/sap/SapODataSource.java:
```diff
@@ -129,86 +152,322 @@ public Schema getSchema() {
+    }
+    switch (propertyMetadata.getType()) {
       case "Binary":
+      case "Edm.Binary":
```
how come there are copies of each of these types?

vladhlinskiy replied:
OData 4 client changed to use XML metadata request instead of metadata request since it's not possible to fetch property annotations in the case of missing vocabulary references: https://issues.apache.org/jira/browse/OLINGO-1130
I faced this issue while testing the plugin with public OData V4 services.

So, currently, we use other model classes that use the "Edm." prefix for OData V4 data types.

albertshau commented on src/main/java/io/cdap/plugin/sap/odata/ODataUtil.java:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  public static EdmExpression.EdmExpressionType typeOf(CsdlDynamicExpression dynamic) {
+    if (dynamic.isPath()) {
```
it seems odd that you would need to do this mapping, there isn't a way to get the type directly from a CsdlDynamicExpression? How are all these isX() methods implemented underneath if they aren't checking some enum?

vladhlinskiy replied:
Currently, we use Csdl model classes that are not checking enum.
For example, the method above [implemented as](https://github.com/apache/olingo-odata4/blob/master/lib/commons-api/src/main/java/org/apache/olingo/commons/api/edm/provider/annotation/CsdlDynamicExpression.java#L205):
```
  /**
   * Returns true if the expression is a edm:Path expression
   * @return true if the expression is a edm:Path expression
   */
  public boolean isPath() {
    return this instanceof CsdlPath;
  }
```
This mapping is added to use `switch-case` instead of duplication of such checks.

albertshau replied:
i see, that's unfortunate

albertshau reviewed (commented):
I'm having trouble following what is happening here. It seems to be mostly logic around translating some annotation representation into the correct StructuredRecord, but I'm confused about what annotations are and how they are represented. For example, are there different types of annotations? Why do we need different schemas?

Can you restructure things so that annotation conversion is located in a single class and add javadocs with example input and what they output? Basically, the next person who touches this code should not have to go through a bunch of SAP docs in order to understand what is happening.

vhlinskyi committed:
CDAP-15761 Address comments

vladhlinskiy commented:
@albertshau, I created `OData4AnnotationToRecordTransformer` so that annotation conversion is located in this class and added Javadocs.
Could you, please, take a look?

albertshau commented on docs/SapEcc-batchsource.md:
```diff
@@ -32,6 +32,85 @@ see [OData URL components].
+  creatable: false
+  updatable: false
+  sortable: false
+  filterable: false
```
shouldn't 'label' be here too?

vladhlinskiy replied:
Added. Thanks!

albertshau commented on docs/SapEcc-batchsource.md:
````diff
@@ -32,6 +32,85 @@ see [OData URL components].
+```
+an entry with property value `true` will be read as the following record:
+```
+value: true
````
where is this coming from?

vladhlinskiy replied:
This is an example of property value. Description changed to:
```
.. with property values "1", "Some Name"(for SoId and BuyerName respectively) will be read as ...

```

albertshau commented on docs/SapEcc-batchsource.md:
````diff
@@ -32,6 +32,85 @@ see [OData URL components].
+  updatable: false
+  sortable: false
+  filterable: false
+```
````
If I understand correctly, if annotations are not included, the output schema would a field named 'BuyerName' that is a non-nullable string. But if annotations are included, the 'BuyerName' field would be a record and not a string? And where the inner 'value' field is the non-nullable string?

I don't think it's a good idea to change the structure of the output based boolean flag. It would be similar to writing a Java method that returns a different object depending on a boolean input argument. This is difficult to explain in documentation and also difficult to use in practice, as the user needs to treat the output differently.

Instead, I think it's better if all of the metadata is added as a separate 'metadata' field

vladhlinskiy replied:
> If I understand correctly, if annotations are not included, the output schema would a field named 'BuyerName' that is a non-nullable string. But if annotations are included, the 'BuyerName' field would be a record and not a string? And where the inner 'value' field is the non-nullable string?

Yes, you are right.

Mapping changed to use separate 'metadata' field.
For example, entity type with properties, annotated as:
```
<EntityType Name="SalesOrder" sap:content-version="1">
  <Key>
    <PropertyRef Name="SoId"/>
  </Key>
  <Property Name="SoId" Type="Edm.String" Nullable="false" MaxLength="10" sap:label="Sales Order ID"
            sap:creatable="false" sap:updatable="false" sap:sortable="false" sap:filterable="false"/>
  <Property Name="BuyerName" Type="Edm.String" Nullable="false" MaxLength="80" sap:label="Full Name"
                    sap:creatable="false" sap:updatable="false" sap:sortable="false" sap:filterable="false"/>
</EntityType>
```
an entry with property values `"1"`, `"Some Name"`(for `SoId` and `BuyerName` respectively) will be read as the following record:
```
SoId: "1"
BuyerName: "Some Name"
metadata:
  SoId:
    label: "Sales Order ID"
    creatable: false
    updatable: false
    sortable: false
    filterable: false
  BuyerName:
    label: "Full Name"
    creatable: false
    updatable: false
    sortable: false
    filterable: false
```

albertshau commented on docs/SapEcc-batchsource.md:
```diff
@@ -32,6 +32,85 @@ see [OData URL components].
+- value - for OData property value.
+- metadata-annotations - metadata annotations record.
+
+OData V4 metadata annotations mapped to a record with fields that correspond to the property attributes.
```
is this supposed to be V2?

vladhlinskiy replied:
Changed. Thanks!

vhlinskyi committed:
CDAP-15761 Mapping changed to add all of the metadata as a separate 'metadata' field

vhlinskyi committed:
CDAP-15761 Docs updated with new metadata mapping

vladhlinskiy commented:
@albertshau I addressed your comments. Could you please review?

Also, [in the previous PR, I mentioned](https://github.com/data-integrations/sap/pull/4#discussion_r340758905
) that I faced issues with records' names overlapping for geospatial types(if record schema with name 'test' contains a field with a record schema with the same name 'test').

I faced this issue again for complex types and noticed that it may occur even in the case when multiple parent records with different record schema names contain records with the same record schema names.
![duplicate-2](https://user-images.githubusercontent.com/51081745/68943559-2e44f400-07b4-11ea-85da-352b2c80539e.png)

In the example above top-level fields have record schema with names "Address-complex-type-values" and "HomeAddress-complex-type-values" respectively.
The record schema name for "City" fields is "City-complex-type-values" in both cases. This leads to "Error: duplicate type name".
  
I resolved this issue in the latest commit by appending UUIDs to record names. Also, there is an option to construct record names manually and pass them as arguments.
Please let me know if this should be changed.

albertshau commented:
The record schema name is supposed to uniquely identify that record within the schema. Without this, it would be impossible to represent recursive structures (ex: a Node class that has a Node field).

So if we know the schema is always going to be the same, it should be a constant and used everywhere. A City is always a City, regardless of which field it is.

vladhlinskiy commented:
@albertshau the latest commit reverted, could you please take a look? 

Please, note that UI will show "Error: duplicate type name" for an inferred schema that contains records' schemas with the same name.
For example, `GeographyPoint` and `GeometryPoint` both mapped to a record with `"point-record"` name:  
![duplicate-point-type-name](https://user-images.githubusercontent.com/51081745/69054690-d0fbad80-0a14-11ea-9811-3138751afbce.png)

Tested on CDAP, which is built from [92707a](https://github.com/cdapio/cdap/tree/92707a9331d230e890d712e6c00ebb3bdc585ba7).
